src/psyclone/psyir/symbols/datatypes.py

```diff
@@ -362,57 +393,83 @@ def _validate_shape(self, extents):
-                    if not ((symbol.is_scalar and symbol.datatype.intrinsic ==
-                             ScalarType.Intrinsic.INTEGER) or
-                            isinstance(symbol.datatype,
-                                       (UnknownFortranType, DeferredType))):
+                _validate_data_node(dimension)
+            elif isinstance(dimension, tuple):
+                if len(dimension) != 2:
+                    raise TypeError(
+                        "A DataSymbol shape-list element specifying lower "
+                        "and upper bounds must be a 2-tuple but '{0}' has {1} "
+                        "entries.".format(str(dimension), len(dimension)))
+                for dim in dimension:
+                    if isinstance(dim, DataNode):
+                        _validate_data_node(dim)
+                    elif not isinstance(dim, int):
                         raise TypeError(
-                            "If a datasymbol is used as a dimension "
-                            "declaration then it should be a scalar "
-                            "integer or an unknown type, but '{0}' is a "
-                            "'{1}'.".format(symbol.name, symbol.datatype))
-                # TODO #1089 - add check that any References are not to a
-                # local datasymbol that is not constant (as this would have
-                # no value).
+                            "A DataSymbol shape-list element specifying lower "
+                            "and upper bounds must be a 2-tuple containing "
+                            "either int or DataNode entries but '{0}' contains"
+                            " '{1}'".format(str(dimension),
+                                            type(dim).__name__))
             elif not isinstance(dimension, (self.Extent, int)):
                 raise TypeError(
-                    "DataSymbol shape list elements can only be "
-                    "'int', ArrayType.Extent or 'DataNode' but found "
+                    "DataSymbol shape list elements can only be 'int', "
+                    "ArrayType.Extent, 'DataNode' or tuple but found "
                     "'{0}'.".format(type(dimension).__name__))
 
     def __str__(self):
         '''
         :returns: a description of this array datatype.
         :rtype: str
 
-        :raises InternalError: if an unsupported dimensions type is \
-            found.
+        :raises InternalError: if an unsupported dimensions type is found.
 
         '''
-        # This import must be placed here to avoid circular
-        # dependencies.
-        # pylint: disable=import-outside-toplevel
-        from psyclone.psyir.nodes import DataNode
         dims = []
         for dimension in self.shape:
-            if isinstance(dimension, DataNode):
-                dims.append(str(dimension))
+            if isinstance(dimension, ArrayType.ArrayBounds):
+                dims.append("{0}:{1}".format(str(dimension.lower),
```

Comment:
Since these expressions end up all in a single line (often already inside an attribute field of another line in a psyir.view()), would it be worth to to shorten it? e.g. ignoring the lower bound if its Literal(1) or converting Literals nodes to its node.value. Essentially the contrary to the constructor.

What do you think?

Reply:
Yes, that's a good idea and I've done it.